rfc: update: bnorm: single-pass formula for calculating mean and variance (#1065) (#1235)
oneapi-src/oneDNN: rfcs/20210519-single-pass-bnorm/README.md

implementation requires BN to read input tensor 3 times. Proposed change
will decrease it to 2 reads. There is no known generic algorithm to decrease it
to 1 read, although such efficiency is possible for special cases where input
tensor is small enough.

## Proposal
Use the following formula to calculate variance:
```python
Var(X) = E(X^2) - (E(X))^2
```
where X = input tensor, E = average

This formula allows hardware to calculate both mean and variance in single pass
over input tensor.
For reference, the canonical formula is:
```python
Var(X) = E((X - Mean(X))^2)
```
The benefit of new formula is 25-33% performance improvement. BN kernel
is memory-bound, with new formula it will need to read input tensor 2 times
(1: mean&var, 2: normalize) instead of 3 times (1: mean, 2: var, 3: normalize)

This formula will be used whenever BN kernel calculates mean and
variance, which means forward propagation in training. However, with
non-standard combination of flags, oneDNN's BN implementation may be
configured to calculate those two values in inference as well.

## Integration

The feature is intended to be used mostly by framework teams.

### Experimental build

> To guard regular oneDNN users from accuracy issues but provide it to the
> frameworks for experiments a separate experimental build is proposed. No binary

Reviewer: ```suggestion
To provide this feature to framework teams for experimentation without impacting the accuracy of oneDNN for other users, we propose to use a separate experimental build. No binary
```

Author: Rephrased, thanks

distribution is expected for experimental build. Users that
want to try experimental features will have to build oneDNN on their side.
**Once experimental build is obtained all experimental features are enabled**.

> To control features separately environmental variables are introduced for each

Reviewer: ```suggestion
To control features separately environment variables are introduced for each
```

Author: Fixed, thanks!

experimental feature.

- A new build flag `DNNL_EXPERIMENTAL` is introduced, **disabled by default**;
- To opt-out BN single pass an environment variable

> `DNNL_EXPERIMENTAL_BNORM_STATS_ONE_PASS` is introduced;

Reviewer: ```suggestion
`DNNL_EXPERIMENTAL_BNORM_STATS_ONE_PASS=<true*|false>` is introduced;
```

Author: fixed, thanks

errors:
- Sum and SumOfSquares will be stored in fp32 precision, regardless of input
	data type
- Kahan summation algorithm [[#1]][1] will be used

The last line, max( ) is there to make sure rounding errors won't cause
variance to go negative. Further in BN calculations square root is taken
from variance. Negative value would result in NaN.

New formula is considered experimental for now. By default the old one will be

> used. There needs to be a way for user to choose formula.

Reviewer: shall this paragraph go away (you clarified this in the above integration section)

Author: removed, thanks
